### DISCUSSION TITLE
Where can I find the pattern language reference?

### DISCUSSION THREAD

**Discussion by valnaumov** [WerWolv/ImHex#1652] [Pattern Language Q&A] [answered]
This page https://docs.werwolv.net/pattern-language does not provide any helpful info.

> The Pattern Language is a C++ and Rust inspired 

But really, what is it?

**Comment by WerWolv** [comment, marked as answer]
What exactly are you looking for? The site you linked is the main documentation that should provide a reference to get started. There's also all the patterns in the ImHex-Patterns repo that you can reference. 
If you have any specific questions, feel free to ask here or in the discord

**Reply by valnaumov** [reply]
I ended up using ChatGPT for my questions. It knows a little bit about the pattern language. I'll ask on discord if I have more questions, thanks for your work.

**Reply by valnaumov** [reply]
@WerWolv I do have a question though. How do I define a `struct` alias conditionally? Depending on the file contents, I want to use one struct or another down the code.

**Reply by paxcut** [reply]
If I  understand the question correctly then you could write:
```cpp
if (some_condition) {

    struct A {
        u8 a;
    };

} else {

    struct A {
        u16 a;
    };
}

A test @0;
```
but this also works:
```cpp
struct A {
    if(some_condition)
         u8 a;
    else
        u16 a;
};
A test@0;
```
The result is pretty much the same.

**Reply by valnaumov** [reply]
This will work. Thanks, @paxcut. I was hoping to give these A's different names to add some semantics though.

**Reply by paxcut** [reply]
You can give them any name you want as long as it is a valid C identifier (may contain any alphanumeric + `_`,  but cannot start with a number).  That code I posted is just the simplest example I could think of that shows the desired behavior. You don't have to use one o the other, you can use both at the same time for different parts. You can also use namespaces to have some structs available and not others.The structures have a great deal of flexibility and can adapt to any situation I have encountered so far in the year or so I've been using imhex.

**Reply by valnaumov** [reply]
you mentioned there's a Discord group, but I could not find any link to it on github or the docs website. Could you point me to it?

**Reply by jumanji144** [reply]
https://discord.gg/imhex shoud lead you to it

**Comment by paxcut** [comment]
I personally disagree with that assessment.There are a few domain specific languages designed to analyzed binary files (010, kaitai, hexinator ,...) and the information displayed on that page makes me really want to check it out because of the features it may contain which other similar tool (that are not free and open  like imhex) don't have like templates, pointers, match statements, inheritance,....
